### PR TITLE
Add check hook for pgstat_end_function_usage in fmgr

### PR DESCRIPTION
**Pull request by tanscorpio7**
### Description

Introduce function to check if pg_stat entry exists in local cache for given key.
Add a hook to validate if pg_stat entry still exists before calling pgstat_end_function_usage.

### Issues Resolved

[BABEL-4740]

### Sign Off

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>
 
### Check List

- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/utils/fmgr/fmgr.c [hunk 1]
```diff
@@ -869,10 +870,13 @@ fmgr_security_definer(PG_FUNCTION_ARGS)
 		 * We could be calling either a regular or a set-returning function,
 		 * so we have to test to see what finalize flag to use.
 		 */
-		pgstat_end_function_usage(&fcusage,
-								  (fcinfo->resultinfo == NULL ||
-								   !IsA(fcinfo->resultinfo, ReturnSetInfo) ||
-								   ((ReturnSetInfo *) fcinfo->resultinfo)->isDone != ExprMultipleResult));
+		
+		if (!pltsql_pgstat_function_check_hook || ((&fcusage)->fs != NULL
+		     && ((*pltsql_pgstat_function_check_hook)(fcinfo->flinfo->fn_oid))))
```
Comment (tanscorpio7) [inline src/backend/utils/fmgr/fmgr.c]: We could only do the check for procedures, since with functions such a case would never arise except for triggers.

File: src/backend/utils/fmgr/fmgr.c [hunk 2]
```diff
@@ -47,7 +48,8 @@ PGDLLIMPORT fmgr_hook_type fmgr_hook = NULL;
 PGDLLIMPORT non_tsql_proc_entry_hook_type non_tsql_proc_entry_hook = NULL;
 PGDLLIMPORT get_func_language_oids_hook_type get_func_language_oids_hook = NULL;
 PGDLLIMPORT pgstat_function_wrapper_hook_type pgstat_function_wrapper_hook = NULL;
-set_local_schema_for_func_hook_type set_local_schema_for_func_hook;
+pltsql_pgstat_function_check_hook_type pltsql_pgstat_function_check_hook = NULL;
+set_local_schema_for_func_hook_type set_local_schema_for_func_hook = NULL;
```
Comment (surendravishnoi) [inline src/backend/utils/fmgr/fmgr.c:50]: why this change?

File: src/backend/utils/fmgr/fmgr.c [hunk 3]
```diff
@@ -869,10 +871,14 @@ fmgr_security_definer(PG_FUNCTION_ARGS)
...
 		 */
-		pgstat_end_function_usage(&fcusage,
-								  (fcinfo->resultinfo == NULL ||
-								   !IsA(fcinfo->resultinfo, ReturnSetInfo) ||
-								   ((ReturnSetInfo *) fcinfo->resultinfo)->isDone != ExprMultipleResult));
+		
+		if (!pltsql_pgstat_function_check_hook ||
+		    (fcache->prokind != PROKIND_PROCEDURE && !CALLED_AS_TRIGGER(fcinfo)) ||
+		    ((&fcusage)->fs != NULL && (*pltsql_pgstat_function_check_hook)(fcinfo)))
+			pgstat_end_function_usage(&fcusage,
```
Comment (surendravishnoi) [inline src/backend/utils/fmgr/fmgr.c:882]: is it possible to put the hook inside pgstat_end_function_usage sp that we do not need to check it for every caller?

File: src/backend/utils/fmgr/fmgr.c (hunk 2 above)
Comment (tanscorpio7) [reply src/backend/utils/fmgr/fmgr.c:50]: unrelated change. Just for consistency in the file.

File: src/backend/utils/fmgr/fmgr.c (hunk 3 above)
Comment (tanscorpio7) [reply src/backend/utils/fmgr/fmgr.c:882]: pgstat_end_function_usage does not contain enough information for the hook function. We need proc OID, kind, etc which are not available there.

**Review by surendravishnoi — APPROVED** [review]
